File: JUNO.cpp
#include <stdlib.h>
#include <stdio.h>
#define MOD 1000000007

long long int modpow(long long int base, long long int exp, long long int modulus)
{
    base %= modulus;
    long long int result = 1;
    while (exp > 0)
    {
        if (exp & 1) result = (result * base) % modulus;
        base = (base * base) % modulus;
        exp >>= 1;
    }
    return result;
}


int main(){
  long long int t,i,j,k;
  long long int V,N;
  scanf("%lld",&t);
  while(t--){
    long long int P[100];
    long long int Q[100];
    long long int L[100];

    long long int V,N,p0,p1,A0,B0,C0,M0,q0,q1,A1,B1,C1,M1;

    scanf("%lld %lld",&V,&N);
    scanf("%lld %lld %lld %lld %lld %lld",&p0,&p1,&A0,&B0,&C0,&M0);
    scanf("%lld %lld %lld %lld %lld %lld",&q0,&q1,&A1,&B1,&C1,&M1);

    P[0] = p0;
    P[1] = p1;
    Q[0] = q0;
    Q[1] = q1;
    for(j=2;j<N;j++){
      P[j] = ( A0*A0*P[j-1] + B0*P[j-2] + C0 )%M0;
      Q[j] = ( A1*A1*Q[j-1] + B1*Q[j-2] + C1 )%M1;
    }
    for(j=0;j<N;j++){
      L[j] = P[j]*(M1)+ Q[j] + 1;
    }

    long long int ans,temp;
    ans = V;
    for(j=0;j<N;j++){
      ans = (modpow(ans,L[j]-1,MOD))%MOD;
    }
    
    printf("%lld\n",ans);
  }
  return 0;
}

File: juno.py
MOD = 1000000007
 
def modpow(base, exp):
    base %= MOD;
    result = 1;
    while (exp > 0):   
        if (exp & 1):
            result = (result * base) % MOD;
        base = (base * base) % MOD;
        exp >>= 1;    
    return result;
 
def main():
    P = [0]*200;
    Q = [0]*200;
    A = [0]*2;
    B = [0]*2;
    C = [0]*2;
    M = [0]*2;
    for i in range(int(raw_input())):
 
        V,N = map(int,raw_input().split());
        P[0],P[1],A[0],B[0],C[0],M[0] = map(int,raw_input().split());
        Q[0],Q[1],A[1],B[1],C[1],M[1] = map(int,raw_input().split());

        if(V%MOD == 0):
            print "0"
        else:
            ans = 1;
            temp = (P[0]*M[1]+Q[0]);
            if(temp != 0):
                ans *= temp;
                temp = (P[1]*M[1]+Q[1]);
            if(temp != 0):
                ans *= temp;
 
            for i in range(2,N):
                P[i] = ( (A[0]*A[0]*P[i-1]) + ( B[0]*P[i-2] ) + C[0])%M[0];
                Q[i] = ( (A[1]*A[1]*Q[i-1]) + ( B[1]*Q[i-2] ) + C[1])%M[1];
                temp = ( P[i]*M[1] + Q[i])
                if(temp != 0):
                    ans *= temp;
 
            print modpow(V,ans%(MOD-1))
        
main()
